keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1284 @@
+beyond this KEP.
+
+In the case of Pods, kube-apiserver will store the received `reason` in the
+already existing `reason` field in the `PodStatus`:
+```
+// PodStatus represents information about the status of a pod. Status may trail the actual
+// state of a system.
+type PodStatus struct {
+  ...
+	// A brief CamelCase message indicating details about why the pod is in this state. e.g. 'Evicted'
+	// +optional
+	Reason string
+  ...
+}
+```
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled.
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed and terminate all running pods.
+	BackoffActionTerminate BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	BackoffActionIgnore BackoffPolicyAction = "Ignore"
+
+)
+
+// BackoffPolicySource specifies the source of the failure
+type BackoffPolicySource string
+
+const (
+
+	BackoffPolicySourceExitCode BackoffPolicySource = "ExitCode"
+
+	BackoffPolicySourceStatusReason BackoffPolicySource = "StatusReason"
+
+)
+
+type BackoffPolicyRequirementOperator string
+
+const (
+	BackoffPolicyRequirementOperatorIn                      BackoffPolicyRequirementOperator = "In"
+	BackoffPolicyRequirementOperatorNotIn                   BackoffPolicyRequirementOperator = "NotIn"
+  BackoffPolicyRequirementOperatorInClosedRange           BackoffPolicyRequirementOperator = "InClosedRange"
+  BackoffPolicyRequirementOperatorNotInClosedRange        BackoffPolicyRequirementOperator = "NotInClosedRange"
+)
+
+// BackoffPolicy describes how failed pods influence the backoffLimit.
+type BackoffPolicy struct {
+	// A list of backoff policy rules. The rules are evaluated in order.
+	// Once a rule matches a Pod failure, the remaining of the rules are ignored.
+	Rules []BackoffPolicyRule
+}
+
+type BackoffPolicyRule struct {
+	// Specifies the action taken on a pod failure when the requirements are satisfied.
+	Action BackoffPolicyAction
+	// A list of backoff policy requirements. The requirements are ANDed.
+	Requirements []BackoffPolicyRequirement
+}
+
+type BackoffPolicyRequirement struct {
+	// Specifies the source of the failure. Valid sources are:
+	// - ExitCode: The exit code from a container
+	// - StatusReason: The reason for the Pod failed status.
+	Source BackoffPolicySource
+
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name. When empty the rule applies to all containers
+	// +optional
+	ContainerName *string
+
+	// Represents the source’s relationship to a set of values.
+	// Valid Operators are In, NotIn, Gt and Lt.
```

Comment:
Also InClosedRange

Reply:
I no longer enumerate all operators here in the comment - there a list of them above